### PR TITLE
Bugfix in backward of linalg.eigh

### PR DESCRIPTION
**Pull request by podgorskiy**
Summary:
Bugfix in backward of linalg.eigh

The expression `VhgV = VhgV - at::matmul(V.mH(), V * at::real(diag_VhgV).unsqueeze(-2))` was simplified to  `VhgV = 0.5 *(vhgv - vhgv.T)`  for  hermitian matrices, which does not seem to be correct.

I do not understand where `VhgV = 0.5 *(vhgv - vhgv.T)` came from, but the generic variant `VhgV = VhgV - at::matmul(V.mH(), V * at::real(diag_VhgV).unsqueeze(-2))` must work for all cases, and it yields different result.

See https://www.internalfb.com/intern/anp/view?id=6872083 for more detail.

The aforementioned simplification was introduced in  D33530149.

I also do not know why tests are passing with this bug, or how tests work. My speculation is that the bug does not make gradients completely wrong, it only messes up the component due to normalization of eigenvectors to norm 1. Also depending on how exactly tests are run, it is possible that finite differences do not produce adequate numerical derivative because there is no unique eigen value decomposition and variation of input matrix may produce jumps in the output values.

Test Plan:
CI + Test example:

```
# Test data
A = th.as_tensor(
    [
        [0.7124138474464417, 1.7507718801498413, 1.3207206726074219],
        [1.7507718801498413, -0.9052235484123230, 1.3285166025161743],
        [1.3207206726074219, 1.3285166025161743, 0.9041665196418762],
    ]
).requires_grad_()

# Some arbitrary gradients
gL = th.as_tensor([-0.5, -0.4, -0.3])
gV = th.as_tensor(
    [
        [-1.2920498847961426, -0.8850541710853577, -1.0448201894760132],
        [2.084738254547119, -0.3660943806171417, 1.0761417150497437],
        [0.3967193365097046, -0.7547367811203003, 0.9107910990715027],
    ]
)
# Do eigenval decomposition
A.grad = None
L, V = th.linalg.eigh(A)

L.backward(gL, retain_graph=True)
V.backward(gV)
A.grad
```

Currently this produces:

```
tensor([[-0.7728,  0.0490,  0.0705],
        [ 0.0490, -0.3912,  0.0465],
        [ 0.0705,  0.0465, -0.0360]])
```

However the expected values:

```
vhgv = V.T @ gV

m = vhgv - V.T @ (V * th.diag(vhgv)[..., None, :])

F = 1.0/(L[..., None, :] - L[..., None])
F[0,0] = 1 
F[1,1] = 1
F[2,2] = 1

gA = V @ (th.diag(gL) + F * m) @ V.T
print(gA)

tensor([[-0.7728,  0.1188, -0.0907],
        [-0.0208, -0.3912,  0.3343],
        [ 0.2317, -0.2413, -0.0360]], grad_fn=<MmBackward0>)
```

Differential Revision: D72221098

### CODE REVIEW

**Comment by podgorskiy** [comment]
cc @lezcano

**Review by lezcano — CHANGES_REQUESTED** [review]
The gradient always lies in the tangent of the manifold. If the matrix is hermitian positive definite the gradient should be hermitian. If the matrix is generally invertible, the gradient will be arbitrary.

The relation between the two gradients is that the hermitian gradient should be the orthogonal projection of the gradient of the unconstrained matrix.

Even if it looks like `eig` and `eigh` are the same function, they are not. Note that a mathematical function is a map *from a domain* into a codomain. This becomes important in differential geometry in general, where a function and its restriction to a subset may have very different properties, like it is the case here.

**Comment by podgorskiy** [comment]
@lezcano , I see now, thank you! 

So the `eigh` is not just a faster `eig` for hermitian case, but it actually produces gradients that are computed given that the matrix is constraint to be hermitian. Am I right? Perhaps not super obvious detail.

Meaning that if I take the gradient from `eig` and project it onto the manifold  of the hermitian matrices then I will get the same gradient as from `eigh`.

I guess there is still the missing zero-out of the diagonal (see l.3812 in the diff) when `gL` is not defined which is require because of the F_{ii} = 0. For hermitian case this seems to not have implications (since diagonal is zero anyway), but I'm not sure about the non-hermitian case.

**Comment by podgorskiy** [comment]
@lezcano, it would help me greatly if you could provide a hint why `VhgV = 0.5 * (VhgV - VhgV.mH());` in particular makes the gradient be projected onto the manifold of the hermitian matrices.

**Comment by lezcano** [comment]
First off, `1/2*(A-A.mH)` is a projection onto the skew-Hermitian matrices. This happens to be the tangent space to the orthogonal matrices in the real case (the orthogonal group) and the unitary matrices (the unitary group) in the complex case.  (it's not the tangent space but the associated Lie algebra, that is, the tangent space at the identity, but yeah).

As to why it is a projection onto the skew-Hermitian matrices, a projection is a function that's the identity in a given set and zero in the rest. See that any matrix can be decomposed into a symmetric part and an antisymmetric part (for the complex case just change `mT` by `mH` and "symmetric" by "Hermtian") by the decomposition A_{sym} = 1/2*(A+A.mT) A_{skew} = 1/2*(A-A.mT) so that A = A_{sym} + A_{skew}, so we can split the space of all matrices into the (direct) sum of symmetric and skew-symmetric matrices. With this in hand, it should be clear that the formula above:
- Is the identity on skew-symmetric matrices
- It's zero on symmetric matrices
in other words, it's a projection onto the skew-symmetric matrices. You can also prove that's an orthogonal projection looking at the inner product on R^{nxn}.

**Comment by lezcano** [comment]
Closing this as the current formula is correct (as proven by the numerical approximation).
